File: Base/liveness.py
"""This file contains code for Register Liveness analysis and
optimization depending on it, e.g. useless code removal"""

import dataclasses
from typing import List, Tuple, Set, Dict
import enum

from Base import ir
from Base import cfg
from Base import opcode_tab as o


@dataclasses.dataclass()
class Liveness:
    """Liveness summary for Bbl

    Only live_out is used for optimizations. live_in, live_def, live_use
    are used  only during the analysis phase.
    """

    live_in: Set[ir.Reg] = dataclasses.field(default_factory=set)
    live_out: Set[ir.Reg] = dataclasses.field(default_factory=set)
    # defined
    live_def: Set[ir.Reg] = dataclasses.field(default_factory=set)
    # used before defined
    live_use: Set[ir.Reg] = dataclasses.field(default_factory=set)


def InsMaybeReplaceDefReg(ins: ir.Ins, reg_old: ir.Reg, reg_new: ir.Reg) -> int:
    """If the ins writes reg_old, replace it with reg_new """
    if ins.opcode.def_ops_count() == 0:
        return 0
    assert ins.opcode.def_ops_count() == 1
    op = ins.operands[0]
    if op == reg_old:
        ins.operands[0] = reg_new
    return 1


def InsMaybeReplaceUseReg(ins: ir.Ins, reg_old: ir.Reg, reg_new: ir.Reg) -> int:
    """If the ins reads reg_old, replace it with reg_new """
    it = enumerate(ins.operands)
    # skip register writing operands
    for _ in range(ins.opcode.def_ops_count()):
        next(it)
    count = 0
    for n, op in it:
        if op == reg_old:
            ins.operands[n] = reg_new
            count += 1
    return count


def _InsUpdateDefUse(ins: ir.Ins, fun: ir.Fun, defs: Set[ir.Reg], uses: Set[ir.Reg]):
    """Compute the set of defined and used registers for each instruction"""
    if ins.opcode.is_call():
        # note: a call instruction may have at most one used reg if opcode is JSR
        callee: ir.Fun = cfg.InsCallee(ins)
        assert isinstance(callee, ir.Fun)
        for cpu_reg in callee.cpu_live_out:
            # we intentionally iterate overall regs here
            for reg in fun.regs:
                if reg.cpu_reg is cpu_reg:
                    defs.add(reg)
                    uses.discard(reg)
        # for cpu_reg in callee.cpu_live_clobber:
        #     defs.add(cpu_reg)
        #     uses.discard(cpu_reg)
        # for cpu_reg in callee.cpu_live_in:
        #     uses.add(cpu_reg)

    num_defs = ins.opcode.def_ops_count()
    for n, reg in enumerate(ins.operands):
        if not isinstance(reg, ir.Reg): continue
        if n < num_defs:
            defs.add(reg)
            uses.discard(reg)
        else:
            uses.add(reg)


def _BblDefUse(bbl: ir.Bbl, fun: ir.Fun) -> Tuple[Set[ir.Reg], Set[ir.Reg]]:
    bbl_def: Set[ir.Reg] = set()
    bbl_use: Set[ir.Reg] = set()
    for ins in reversed(bbl.inss):
        _InsUpdateDefUse(ins, fun, bbl_def, bbl_use)
    return bbl_def, bbl_use


def _InsUpdateLiveness(ins: ir.Ins, fun: ir.Fun, live_out: Set[ir.Reg]) -> bool:
    """Similar to _InsUpdateDefUse but also checks if the instruction is useless"""
    if ins.opcode.is_call():
        # note: a call instruction may have at most one used reg if opcode is JSR
        callee: ir.Fun = cfg.InsCallee(ins)
        assert isinstance(callee, ir.Fun)
        for cpu_reg in callee.cpu_live_out:
            for reg in fun.regs:
                if reg.cpu_reg is cpu_reg:
                    live_out.discard(reg)

    # TODO: take cpu_live_in into account, otherwise we cannot eliminate useless code
    #       after pusharg and poparg convdersions
    #         live_out.discard(cpu_reg)
    #     for cpu_reg in callee.cpu_live_clobber:
    #         live_out.discard(cpu_reg)
    #     for cpu_reg in callee.cpu_live_in:
    #         live_out.add(cpu_reg)

    is_live = ins.opcode.has_side_effect()
    num_defs = ins.opcode.def_ops_count()
    for n, reg in enumerate(ins.operands):
        if not isinstance(reg, ir.Reg): continue

        if n < num_defs:
            if reg in live_out:
                is_live = True
            live_out.discard(reg)
        else:
            # all defs precede the uses, so if the ins is not live at this point,
            # we can skip the
            if not is_live:
                break
            live_out.add(reg)
    return is_live


def _BblRemoveUselessInstructions(bbl: ir.Bbl, fun: ir.Fun) -> int:
    live_out = bbl.live_out.copy()
    old_count = len(bbl.inss)
    keep = []
    for ins in reversed(bbl.inss):
        if _InsUpdateLiveness(ins, fun, live_out):
            keep.append(ins)
    bbl.inss = list(reversed(keep))
    return old_count - len(keep)


def FunRemoveUselessInstructions(fun: ir.Fun) -> int:
    assert ir.FUN_FLAG.LIVENESS_VALID in fun.flags
    return ir.FunGenericRewriteBbl(fun, _BblRemoveUselessInstructions)


def _FunLivenessFixpoint(fun: ir.Fun,
                         all_liveness: Dict[str, Liveness]) -> int:
    """ Standard backward flow liveness computation"""
    count = 0
    active = fun.bbls[:]
    while active:
        count += 1
        bbl = active.pop(-1)
        # print(f"@@ processing {bbl.name}")
        liveness = all_liveness[bbl.name]
        live_in = liveness.live_out.copy()
        live_in.difference_update(liveness.live_def)
        live_in.update(liveness.live_use)
        if len(live_in) <= len(liveness.live_in):
            continue
        liveness.live_in = live_in

        for pred in bbl.edge_in:
            pred_liveness = all_liveness[pred.name]
            old_len = len(pred_liveness.live_out)
            pred_liveness.live_out.update(liveness.live_in)
            if len(pred_liveness.live_out) > old_len:
                if pred not in active:
                    active.append(pred)

    return count


def FunComputeLivenessInfo(fun: ir.Fun) -> int:
    """Assumes that cfg.funInitCFG has been called"""
    if len(fun.bbls) > 1:
        assert len(fun.bbls[0].edge_out) > 0, f"you must run cfg.FunInitCFG"
    all_liveness: Dict[str, Liveness] = {}
    for bbl in fun.bbls:
        liveness = Liveness()
        liveness.live_def, liveness.live_use = _BblDefUse(bbl, fun)
        # if bbl.IsReturn():
        #     liveness.live_out = set(fun.cpu_live_out)
        all_liveness[bbl.name] = liveness
    rounds = _FunLivenessFixpoint(fun, all_liveness)
    for bbl in fun.bbls:
        bbl.live_out = all_liveness[bbl.name].live_out
    fun.flags |= ir.FUN_FLAG.LIVENESS_VALID
    return rounds


def _HandleSpillForIns(ins: ir.Ins, regs_to_be_spilled: Set[str],
                       spill_slots, ld_spill, st_spill, zero):
    out_ld = []
    out_st = []

    def_count = ins.opcode.def_ops_count()

    def maybe_add_spill(pos, reg: ir.Reg):
        if reg.name not in regs_to_be_spilled:
            return
        if pos < def_count:
            out_st.append(ir.Ins(
                st_spill, [spill_slots[reg.name], zero, reg]))
        else:
            out_ld.append(ir.Ins(
                ld_spill, [reg, spill_slots[reg.name], zero]))

    for n, op in enumerate(ins.operands):
        if isinstance(op, ir.Reg):
            maybe_add_spill(n, op)

    return out_ld + [ins] + out_st


BEFORE_BBL = -32000
AFTER_BBL = 32000
NO_USE = AFTER_BBL + 1


@enum.unique
class LiveRangeFlag(enum.Flag):
    LAC = 1  # live across call
    PRE_ALLOC = 2  # already allocated - re-use reg after last_use
    IGNORE = 4  # ignore completely


@dataclasses.dataclass()
class LiveRange:
    """Represents and intra Bbl live-range

    if reg == REG_INVALID - this is a fake liverange
    """
    def_pos: int
    last_use_pos: int
    reg: ir.Reg  # contains a proper register
    num_uses: int
    uses: List["LiveRange"] = dataclasses.field(default_factory=list)
    flags: LiveRangeFlag = LiveRangeFlag(0)
    cpu_reg: ir.CpuReg = ir.CPU_REG_INVALID

    def is_cross_bbl(self):
        return self.last_use_pos is AFTER_BBL or self.def_pos is BEFORE_BBL

    def is_use_lr(self):
        return self.reg is ir.REG_INVALID

    def __lt__(self, other: "LiveRange"):
        """This will order uses before defs

        because def_pos and last_use_pos are the same for "use lr's" but different for def lr's
        """
        return ((self.def_pos, self.last_use_pos) <
                (other.def_pos, other.last_use_pos))

    def __repr__(self):
        def render_pos(pos):
            if pos == BEFORE_BBL: return "BB"
            if pos == AFTER_BBL: return "AB"
            if pos == NO_USE: return "NU"
            return f"{pos:2d}"

        flags_str = ""
        if self.flags:
            flags_str = f" {' '.join(f.name for f in LiveRangeFlag if f in self.flags)}"

        if self.is_use_lr():
            # commented to make output compatible with c++ implementation
            starts = ",".join([f"{lr.reg.name}:{lr.def_pos}" for lr in self.uses])
            extra_str = f" uses:{starts}"
            # extra_str = f" uses:{len(self.uses)}"
        else:
            extra_str = f" def:{self.reg.name}:{self.reg.kind.name}"
            if self.cpu_reg is ir.CPU_REG_SPILL:
                flags_str += " SPILLED"
            elif self.cpu_reg != ir.CPU_REG_INVALID:
                extra_str += "@" + self.cpu_reg.name
        return f"LR {render_pos(self.def_pos)} - {render_pos(self.last_use_pos)}{flags_str}{extra_str}"


def BblGetLiveRanges(bbl: ir.Bbl, fun: ir.Fun, live_out: Set[ir.Reg], emit_uses: bool) -> List[LiveRange]:
    """ LiveRanges are use to do register allocation

    Note: function call handling is quite adhoc and likely has bugs.
    The output contains the following special LiveRanges
    * LRs without a last_use if the register is used outside the Bbl (based on live_out)
    * LRs without a def of the register is defined outside the Bbl
    * [if emit_uses] use-def LRs contain the LRs of all the used regs in the instruction at point p.
                     (def=p last_use=p,  reg=REG_INVALID)
    """
    out = []
    bbl_size = len(bbl.inss)

    last_use: Dict[ir.Reg, LiveRange] = {}
    last_call_pos = -1
    last_call_cpu_live_in = []

    def initialize_lr(pos: int, reg: ir.Reg) -> LiveRange:
        lr = LiveRange(-1, pos, reg, 1)
        last_use[reg] = lr
        out.append(lr)
        return lr

    def finalize_lr(lr: LiveRange, def_pos: int):
        lr.def_pos = def_pos
        if (last_call_pos != -1 and last_call_pos != AFTER_BBL and
                last_call_pos < lr.last_use_pos):
            lr.flags |= LiveRangeFlag.LAC
        del last_use[lr.reg]

    # handle live ranges that extend passed the bbl
    for reg in live_out:
        initialize_lr(AFTER_BBL, reg)

    for pos, ins in enumerate(reversed(bbl.inss)):
        pos = bbl_size - 1 - pos
        if ins.opcode is o.RET:
            if fun.cpu_live_out:
                last_call_cpu_live_in = fun.cpu_live_out
                last_call_pos = AFTER_BBL
        elif ins.opcode.is_call():
            callee: ir.Fun = cfg.InsCallee(ins)
            assert isinstance(callee, ir.Fun)
            # This complication only applies after we have (partial) reg allocation
            # Finalize live ranges using the results of the call
            if callee.cpu_live_out:
                # Note, destructive list iteration -> `list(...)` is necessary
                for reg, lr in list(last_use.items()):
                    if reg.HasCpuReg() and reg.cpu_reg in callee.cpu_live_out:
                        finalize_lr(lr, pos)
            last_call_cpu_live_in = callee.cpu_live_in
            last_call_pos = pos  # setting this after dealing with cpu_live_out seems right

        num_defs = ins.opcode.def_ops_count()
        uses = []
        for n, reg in enumerate(ins.operands):
            if not isinstance(reg, ir.Reg): continue
            if n < num_defs:  # define reg
                lr = last_use.get(reg)
                if lr:
                    finalize_lr(lr, pos)
                else:
                    last_use_pos = NO_USE
                    # Note: likely this makes some assumptions about the adjacency
                    # of these instruction and the call: We assume this cannot be LAC!
                    if reg.HasCpuReg() and reg.cpu_reg in last_call_cpu_live_in:
                        last_use_pos = last_call_pos
                    out.append(LiveRange(pos, last_use_pos, reg, 0))
            else:  # used reg
                lr = last_use.get(reg)
                if lr:
                    lr.num_uses += 1
                else:
                    lr = initialize_lr(pos, reg)
                if lr not in uses:
                    uses.append(lr)
        if emit_uses and uses:
            # Note "pos, pos" ensure that this record will come before
            #       a regular record after sorting
            out.append(LiveRange(pos, pos, ir.REG_INVALID, 0, uses))

    for lr in list(last_use.values()):
        finalize_lr(lr, BEFORE_BBL)
    return out

File: Base/liveness.cc
// (c) Robert Muth - see LICENSE for more info

#include "Base/liveness.h"
#include "Base/cfg.h"
#include "Base/ir.h"
#include "Base/serialize.h"

#include <algorithm>
#include <iomanip>
#include <set>

namespace cwerg::base {
namespace {

#if 0
void DumpLiveness(BitVec bv, Fun fun, std::ostream* output) {
  const unsigned num_regs = FunNumRegs(fun);
  const HandleVec reg_names = FunRegNames(fun);
  for (unsigned i = 0; i < num_regs; ++i) {
    if (!bv.BitGet(i)) continue;
    *output << " " <<  Str(reg_names.Get(i));
  }
}
#endif

int FunLivenessFixpoint(Fun fun, BitVec live, BitVec old) {
  // Note we look at the last bbl first
  int count = 0;
  std::set<Bbl> active_set;
  std::vector<Bbl> active_stk;
  for (Bbl bbl : FunBblIter(fun)) {
    active_stk.push_back(bbl);
    active_set.insert(bbl);
  }

  while (!active_stk.empty()) {
    ++count;
    const Bbl bbl = active_stk.back();
    active_stk.pop_back();
    active_set.erase(bbl);
#if 0
    std::cout << "fixpoint BBL " << Name(bbl) << " out: ";
    DumpLiveness(BblLiveOut(bbl), fun, &std::cout);
    std::cout << "\n";
#endif
    live.CopyFrom(BblLiveOut(bbl));
    live.AndNotWith(BblLiveDef(bbl));
    live.OrWith(BblLiveUse(bbl));
    if (live.Equal(BblLiveIn(bbl))) continue;
    /*
    std::cout << Name(bbl) << ": ";
    RenderRegBitVec(fun, BblLiveOut(bbl), &std::cout);
    std::cout <<  " -> ";
    RenderRegBitVec(fun, live, &std::cout);
    std::cout <<  "\n";
  */
    for (Edg edg : BblPredEdgIter(bbl)) {
      Bbl pred = EdgPredBbl(edg);
      old.CopyFrom(BblLiveOut(pred));
      BblLiveOut(pred).OrWith(live);
      if (old.Equal(BblLiveOut(pred))) continue;
      // NOTE: would it be better to go DFS and move the succ up the stack?
      if (active_set.find(pred) == active_set.end()) active_stk.push_back(pred);
    }
  }
  return count;
}

void InsUpdateDefUse(Ins ins,
                     BitVec defs,
                     BitVec uses,
                     unsigned num_regs,
                     const Reg reg_map[]) {
  if (InsOpcode(ins).IsCall()) {
    Fun callee = InsCallee(ins);
    for (unsigned j = 0; j < FunNumCpuLiveOut(callee); ++j) {
      const CpuReg cpu_reg = FunCpuLiveOut(callee)[j];
      // terminate live ranges for callee result regs
      for (unsigned i = 0; i < num_regs; ++i) {
        const Reg reg = reg_map[i];
        if (RegCpuReg(reg) == cpu_reg) {
          defs.BitSet(RegNo(reg));
          uses.BitClear(RegNo(reg));
        }
      }
    }
  }

  // for function calls deal with clobbering machine regs
  const unsigned num_ops = InsOpcode(ins).num_operands;
  const unsigned num_defs = InsOpcode(ins).num_defs;
  for (unsigned i = 0; i < num_defs; ++i) {
    Reg reg = Reg(InsOperand(ins, i));
    defs.BitSet(RegNo(reg));
    uses.BitClear(RegNo(reg));
  }
  for (unsigned i = num_defs; i < num_ops; ++i) {
    const Reg reg = Reg(InsOperand(ins, i));
    if (reg.kind() != RefKind::REG) continue;
    uses.BitSet(RegNo(reg));
  }
}

}  // namespace

bool InsWritesRegInList(Ins ins, BitVec regs) {
  const unsigned num_defs = InsOpcode(ins).num_defs;
  for (unsigned i = 0; i < num_defs; ++i) {
    Reg reg = Reg(InsOperand(ins, i));
    if (regs.BitGet(RegNo(reg))) return true;
  }
  return false;
}

void InsUpdateLiveness(Ins ins,
                       unsigned num_regs,
                       const Reg reg_map[],
                       BitVec live) {
  if (InsOpcode(ins).IsCall()) {
    Fun callee = InsCallee(ins);
    for (unsigned j = 0; j < FunNumCpuLiveOut(callee); ++j) {
      const CpuReg cpu_reg = FunCpuLiveOut(callee)[j];
      // terminate live ranges for callee result regs
      for (unsigned i = 0; i < num_regs; ++i) {
        const Reg reg = reg_map[i];
        if (RegCpuReg(reg) == cpu_reg) {
          live.BitClear(RegNo(reg));
        }
      }
    }
  }
  // TODO: this needs to take FunCpuLiveIn into account
  // for function calls deal with clobbering machine regs
  const unsigned num_defs = InsOpcode(ins).num_defs;
  const unsigned num_ops = InsOpcode(ins).num_operands;
  for (unsigned i = 0; i < num_defs; ++i) {
    const Reg reg = Reg(InsOperand(ins, i));
    live.BitClear(RegNo(reg));
  }
  for (unsigned i = num_defs; i < num_ops; ++i) {
    const Reg reg = Reg(InsOperand(ins, i));
    if (reg.kind() != RefKind::REG) continue;
    live.BitSet(RegNo(reg));
  }
}

int FunRemoveUselessInstructions(Fun fun, std::vector<Ins>* to_delete) {
  to_delete->clear();
  const uint16_t num_regs = FunNumRegs(fun);
  const Reg* const reg_map = (Reg*)FunRegMap(fun).BackingStorage();
  BitVec live = BitVec::New(num_regs);
  std::vector<Ins> useless;

  for (Bbl bbl : FunBblIter(fun)) {
    live.CopyFrom(BblLiveOut(bbl));
    for (Ins ins : BblInsIterReverse(bbl)) {
      // InsRenderToAsm(ins, &std::cout);
      // std::cout << "\n";
      if (!InsOpcode(ins).HasSideEffect() && !InsWritesRegInList(ins, live)) {
        to_delete->push_back(ins);
        continue;
      }
      InsUpdateLiveness(ins, num_regs, reg_map, live);
    }
  }
  for (Ins ins : *to_delete) {
    BblInsUnlink(ins);
    InsDel(ins);
  }
  BitVec::Del(live);
  return to_delete->size();
}

void FunNumberReg(Fun fun) {
  unsigned num_regs = 1;  // do not use zero
  for (Reg reg : FunRegIter(fun)) {
    RegNo(reg) = num_regs++;
  }
  FunNumRegs(fun) = num_regs;
  HandleVec::Del(FunRegMap(fun));
  FunRegMap(fun) = HandleVec::New(num_regs);

  num_regs = 1;
  for (Reg reg : FunRegIter(fun)) {
    FunRegMap(fun).Set(num_regs++, reg);
  }
}

// Returns the number iterations the fixed point computation took
// (Re)Initializes the following fields:
// * RegLiveness
// * FunLiveXXX
// * BblLiveXXX
int FunComputeLivenessInfo(Fun fun) {
  const unsigned num_regs = FunNumRegs(fun);
  const Reg* const reg_map = (Reg*)FunRegMap(fun).BackingStorage();
#if 0
  BitVec::Realloc(&FunLiveOut(fun), num_regs);
#endif

  // Mimic what we do for allocation
  for (Bbl bbl : FunBblIter(fun)) {
    BitVec::Del(BblLiveOut(bbl));
  }
  for (Bbl bbl : FunBblIter(fun)) {
    BitVec::Del(BblLiveIn(bbl));
    BitVec::Del(BblLiveUse(bbl));
    BitVec::Del(BblLiveDef(bbl));
  }

  // We want all the BblLiveOuts to be adjacent
  for (Bbl bbl : FunBblIter(fun)) {
    BblLiveOut(bbl) = BitVec::New(num_regs);
  }
  for (Bbl bbl : FunBblIter(fun)) {
    BblLiveIn(bbl) = BitVec::New(num_regs);
    BblLiveUse(bbl) = BitVec::New(num_regs);
    BblLiveDef(bbl) = BitVec::New(num_regs);
  }

  for (Bbl bbl : FunBblIter(fun)) {
    // std::cout  << "BBL " << Name(bbl) << "\n";
    for (Ins ins : BblInsIterReverse(bbl)) {
      InsUpdateDefUse(ins, BblLiveDef(bbl), BblLiveUse(bbl), num_regs, reg_map);
    }
/*
    std::cout  << "BBL " << Name(bbl) << "  use";
    RenderRegBitVec(fun, BblLiveUse(bbl), &std::cout);
    std::cout  << "  def";
    RenderRegBitVec(fun, BblLiveDef(bbl), &std::cout);
    std::cout  << "\n";
*/
#if 0
    std::cout << "def:";
    DumpLiveness(BblLiveDef(bbl), fun, &std::cout);
    std::cout << "\n";

    std::cout << "use:";
    DumpLiveness(BblLiveUse(bbl), fun, &std::cout);
    std::cout << "\n";
#endif
#if 0
    Ins tail = BblInsList::Tail(bbl);
    if (!tail.isnull() && InsOpcodeKind(tail) == OF::RET) {
      // This is only relevant once we have CPU regs
      // otherwise the push instructions will take care of this
       BblLiveOut(bbl).CopyFrom(FunLiveOut(fun));
    }
#endif
  }

  BitVec use = BitVec::New(num_regs);
  BitVec def = BitVec::New(num_regs);
  int count = FunLivenessFixpoint(fun, use, def);
  BitVec::Del(use);
  BitVec::Del(def);
  return count;
}

bool ListContainsCpuReg(const CpuReg* list, unsigned num_regs, CpuReg cpu_reg) {
  for (unsigned i = 0; i < num_regs; ++i) {
    if (list[i] == cpu_reg) return true;
  }
  return false;
}

std::vector<LiveRange> BblGetLiveRanges(Bbl bbl,
                                        Fun fun,
                                        const std::vector<Reg>& live_out,
                                        bool emit_use_def) {
  unsigned bbl_size = 0;
  for (Ins ins : BblInsIter(bbl)) {
    std::ignore = ins;
    ++bbl_size;
  }

  int16_t last_call_pos = -1;
  const CpuReg* last_call_cpu_live_in = nullptr;
  unsigned last_call_num_cpu_live_in = 0;

  std::vector<LiveRange> out;
  // Add dummy entry at index 0 - so we can treat this as invalid
  out.emplace_back(LiveRange{BEFORE_BBL, BEFORE_BBL, Reg(0), 0, {0, 0, 0, 0}});

  auto initialize_lr = [&](int16_t pos, Reg reg) -> uint16_t {
    uint16_t lr_index = out.size();
    RegLastUse(reg) = lr_index;
    out.emplace_back(LiveRange{-1, pos, reg, 1, {0, 0, 0, 0}});
    return lr_index;
  };

  auto finalize_lr = [&](int16_t lr_no, int16_t def_pos) {
    LiveRange& lr = out[lr_no];
    // BblRenderToAsm(bbl, fun, &std::cout);
    ASSERT(lr.def_pos == -1, "in " << Name(fun) << "unexpected def_pos " << lr);
    lr.def_pos = def_pos;
    if (last_call_pos != -1 && last_call_pos != AFTER_BBL &&
        last_call_pos < lr.last_use_pos) {
      lr.SetFlag(LR_FLAG::LAC);
    }
    RegLastUse(lr.reg) = 0;  // invalid
  };

  for (Reg reg : live_out) {
    initialize_lr(AFTER_BBL, reg);
  }

  int16_t pos = bbl_size;
  for (Ins ins : BblInsIterReverse(bbl)) {
    --pos;
    if (InsOPC(ins) == OPC::RET) {
      if (FunNumCpuLiveOut(fun) > 0) {
        last_call_num_cpu_live_in = FunNumCpuLiveOut(fun);
        last_call_cpu_live_in = FunCpuLiveOut(fun);
        last_call_pos = AFTER_BBL;
      }
    } else if (InsOpcode(ins).IsCall()) {
      Fun callee = InsCallee(ins);
      // This complication only applies after we have (partial) reg allocation
      if (FunNumCpuLiveOut(callee) > 0) {
        // terminate live ranges for callee result regs
        for (unsigned i = 0; i < out.size(); ++i) {
          if (out[i].def_pos == -1) {
            const CpuReg cpu_reg(RegCpuReg(out[i].reg));
            if (cpu_reg.kind() != RefKind::CPU_REG) continue;
            if (ListContainsCpuReg(FunCpuLiveOut(callee),
                                   FunNumCpuLiveOut(callee), cpu_reg)) {
              finalize_lr(i, pos);
            }
          }
        }
      }
      last_call_pos = pos;
      last_call_cpu_live_in = FunCpuLiveIn(callee);
      last_call_num_cpu_live_in = FunNumCpuLiveIn(callee);
    }
    const unsigned num_defs = InsOpcode(ins).num_defs;
    const unsigned num_ops = InsOpcode(ins).num_operands;
    // process definitions
    for (unsigned i = 0; i < num_defs; ++i) {
      const Reg reg(InsOperand(ins, i));
       const CpuReg cpu_reg(RegCpuReg(reg));
      // skip spilled regs
      if (cpu_reg.kind() == RefKind::STACK_SLOT) continue;
      if (i == 0 && RegHasFlag(reg, REG_FLAG::TWO_ADDRESS) && reg == InsOperand(ins, 1)) continue;
      if (RegLastUse(reg) != 0) {
        finalize_lr(RegLastUse(reg), pos);
      } else {
        int16_t last_use_pos = NO_USE;
        if (cpu_reg.kind() == RefKind::CPU_REG &&
            ListContainsCpuReg(last_call_cpu_live_in, last_call_num_cpu_live_in,
                               cpu_reg)) {
          last_use_pos = last_call_pos;
        }
        out.emplace_back(LiveRange{pos, last_use_pos, reg, 0, {0, 0, 0, 0}});
      }
    }

    // process uses
    uint16_t ud[4] = {0, 0, 0, 0}; // use-def links
    unsigned num_ud = 0;
    for (unsigned i = num_defs; i < num_ops; ++i) {
      const Reg reg = Reg(InsOperand(ins, i));
      if (reg.kind() != RefKind::REG) continue;
      const CpuReg cpu_reg(RegCpuReg(reg));
      // skip spilled regs
      if (cpu_reg.kind() == RefKind::STACK_SLOT) continue;
      uint16_t lr_index = RegLastUse(reg);
      if (lr_index == 0) {
        // this is the end of a live range
        lr_index = initialize_lr(pos, reg);
      } else {
        ++out[lr_index].num_uses;
      }
      // record uses but avoid dups
      unsigned pos;
      for (pos = 0; pos < num_ud; ++pos) {
        if (ud[pos] == lr_index) break;
      }
      if (pos == num_ud) {
        ud[num_ud] = lr_index;
        ++num_ud;
      }
    }
    if (emit_use_def && num_ud > 0) {
      out.emplace_back(
          LiveRange{pos, pos, Reg(0), 0, {ud[0], ud[1], ud[2], ud[3]}});
    }
  }

  for (unsigned i = 0; i < out.size(); ++i) {
    if (out[i].def_pos == -1) {
      finalize_lr(i, BEFORE_BBL);
      RegLastUse(out[i].reg) = 0;
    }
  }
  return out;
}

std::ostream& operator<<(std::ostream& os, const LiveRange& lr) {
  auto render_pos = [&os](int16_t pos) {
    if (pos == BEFORE_BBL)
      os << "BB";
    else if (pos == AFTER_BBL)
      os << "AB";
    else if (pos == NO_USE)
      os << "NU";
    else
      os << std::setw(2) << pos;
  };
  os << "LR ";
  render_pos(lr.def_pos);
  os << " - ";
  render_pos(lr.last_use_pos);

  if (lr.HasFlag(LR_FLAG::PRE_ALLOC)) os << " PRE_ALLOC";
  if (lr.HasFlag(LR_FLAG::LAC)) os << " LAC";
  if (lr.HasFlag(LR_FLAG::IGNORE)) os << " IGNORE";
  if (lr.is_use_lr()) {
    unsigned int n = 0;
    for (unsigned i = 0; i < MAX_USES_PER_OPCODE; ++i) {
      if (lr.use_def[i] != 0) ++n;
    }
    std::cout << " uses:" << n;

  } else {
    if (lr.cpu_reg == CPU_REG_SPILL) std::cout << " SPILLED";
    os << " def:" << Name(lr.reg) << ":" << EnumToString(RegKind(lr.reg));
    if (!lr.cpu_reg.isnull()) {
      os << "@" << Name(lr.cpu_reg);
    }
  }
  return os;
}

}  // namespace cwerg::base
